### PR TITLE
Historical rates support

### PR DESCRIPTION
**Pull request by bak1an**
This adds support for fetching all historical rates from http://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist.xml instead of just the last 90 days in `update_historical_rates` method (90 days historical rates is still the default).

Also adds `save_historical_rates` method which calls `save_rates` with proper historical urls.

See the very first commit to get an idea what the actual change was, I tried to make it compatible with existing API not to change any behavior for existing applications.

The rest of commits here are to fix problems that appeared after, namely:

- All historical data contains some now not existing currencies which make money gem throw us exception, I decided to just skip those rates.

- I had to rewrite parser to be SAX parser because working with dom/xpath was too slow on that big file:

```bash
# saving full history to local file:
$ bundle exec ruby -e "require 'eu_central_bank'; EuCentralBank.new.save_historical_rates('/tmp/all.xml', true)" 

# dom/xpath parser, just loading the rates from local file:
$ time bundle exec ruby -e "require 'eu_central_bank'; EuCentralBank.new.update_historical_rates('/tmp/all.xml', true)" 
bundle exec ruby -e   408.07s user 5.07s system 90% cpu 7:34.00 total

# new sax parser, you can see that it is 400+ seconds vs 3 seconds issue:
$ time bundle exec ruby -e "require 'eu_central_bank'; EuCentralBank.new.update_historical_rates('/tmp/all.xml', true)"
bundle exec ruby -e   3.15s user 0.20s system 96% cpu 3.481 total
```

- Bump nokogiri gem version, just because they have some bugfixes.

This looks ready for some initial review, will add specs for new behavior meanwhile.

Refs: #52

### CODE REVIEW

File: lib/eu_central_bank.rb
Comment (antstorm) [inline lib/eu_central_bank.rb]: nit: can you please move this class into a separate file?

**Comment by bak1an** [comment]
@antstorm Looks like this slipped through my fingers back then. Rebased it with:
- Moving SAX document into a separate file
- Workarounding nokogiri jruby SAX bug that was crashing the build (https://github.com/RubyMoney/eu_central_bank/pull/81/commits/bddfb875cf9489b276018dc9d4a146b57dba682d)
- Adding some specs for fetching full history (mostly checking that it will use a different URL when `all` parameter is `true`)

Let me know what you think.

Thanks! (and sorry for the delay)

**Comment by dv** [comment]
@antstorm any chance this can get merged? Seems like a nice upgrade, useful for any accounting software.

**Comment by dv** [comment]
@bak1an I did some rebasing in https://github.com/dv/eu_central_bank/tree/hist-rates feel free to update your branch with this.

Currently your branch does not work with an updated Money gem due to the `mutex`/transaction difference (getting a "cannot call `synchronize` on `nil`" error).

File: lib/ecb/rates_document.rb
Comment (antstorm) [inline lib/ecb/rates_document.rb]: nit: Can you please use the full name here to avoid potential collisions with ECB abbreviation?

**Review by antstorm — DISMISSED** [review]
@bak1an can you please rename the module and rebase it with `master`? Happy to merge otherwise, good job!

**Comment by bak1an** [comment]
Sure, can do it next week.

**Comment by ivan-uqido** [comment]
@bak1an pleeeeease :D

File: lib/ecb/rates_document.rb
Comment (bak1an) [reply lib/ecb/rates_document.rb]: @antstorm Could not come up with anything better than `EuCentralBank` for the module name so moved `RatesDocument` it to be inside that same class. Didn't want to move the class around as it is already used as is everywhere.

**Comment by bak1an** [comment]
@antstorm Rebased and it works. Had to keep jruby nokogiri workaround though, one we have in gemspec is still affected by that issue.

**Comment by ivan-uqido** [comment]
@bak1an @antstorm can't wait :D

**Comment by ivan-uqido** [comment]
?

**Comment by gildesmarais** [comment]
Yesterday, semmons99 closed a bunch of PRs and issues without any further comment. This one is affected, too. Am I missing an spring clean announcement or something? 
This PR looked fine to me and would have added a nice feature to the gem.

**Comment by semmons99** [comment]
closing of long-standing PRs without resolution happy to reopen and merge is @antstorm gives the 👍

**Comment by bak1an** [comment]
@antstorm Any chance we can reopen and merge this? This is still useful (I even run it in production)

**Comment by bak1an** [comment]
Or maybe @semmons99 can point to some other maintainers that are active recently to consider reopening this

**Comment by semmons99** [comment]
@bak1an give me a signal if/when the PR goes green and I will merge and release

**Comment by bak1an** [comment]
@semmons99 merged in `main` branch, need approval for CI to run

**Review by semmons99 — APPROVED** [review]

**Comment by semmons99** [comment]
@bak1an going to merge into main and release now

**Comment by semmons99** [comment]
@bak1an released https://rubygems.org/gems/eu_central_bank/versions/1.7.0

**Comment by bak1an** [comment]
@semmons99 The most awesome, thanks!

**Pull request merged**
